File: src/rate_limit_patterns/backend/lua/leaky_bucket.lua
-- Leaky Bucket Rate Limiter
-- KEYS[1] = rate limit key
-- ARGV[1] = capacity (queue max size)
-- ARGV[2] = leak_rate (items leaked per second)
-- ARGV[3] = current_time (-1 to use Redis TIME)

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local leak_rate = tonumber(ARGV[2])
local supplied_time = tonumber(ARGV[3])

local now = 0

if supplied_time and supplied_time >= 0 then
    now = supplied_time
else
    local time_data = redis.call('TIME')
    now = tonumber(time_data[1]) + (tonumber(time_data[2]) / 1000000)
end

local state = redis.call('HMGET', key, 'queue_size', 'last_leak')
local queue_size = tonumber(state[1]) or 0
local last_leak = tonumber(state[2]) or now

local elapsed = now - last_leak
if elapsed < 0 then
    elapsed = 0
end

local drained = elapsed * leak_rate
local queue_after_leak = queue_size - drained
if queue_after_leak < 0 then
    queue_after_leak = 0
end

local allowed = 0
local retry_after = 0
local new_queue_size = queue_after_leak

if queue_after_leak + 1 <= capacity then
    allowed = 1
    new_queue_size = queue_after_leak + 1
else
    if leak_rate > 0 then
        retry_after = math.ceil(((queue_after_leak + 1) - capacity) / leak_rate)
        if retry_after < 1 then
            retry_after = 1
        end
    else
        retry_after = 1
    end
end

local remaining = math.floor(capacity - new_queue_size)
if remaining < 0 then
    remaining = 0
end

redis.call('HMSET', key, 'queue_size', new_queue_size, 'last_leak', now)

local ttl = 1
if leak_rate > 0 then
    ttl = math.ceil(capacity / leak_rate)
    if ttl < 1 then
        ttl = 1
    end
end
redis.call('EXPIRE', key, ttl)

local reset_at = now
if leak_rate > 0 then
    reset_at = now + (new_queue_size / leak_rate)
end

local request_count = math.floor(new_queue_size)

return {allowed, remaining, retry_after, reset_at, request_count}

File: src/rate_limit_patterns/algorithms/leaky_bucket.py
"""Leaky Bucket rate limiting algorithm."""

from __future__ import annotations

import math
from typing import Any

from rate_limit_patterns.models import RateLimitConfig


class LeakyBucketAlgorithm:
    """Leaky Bucket algorithm implementation.

    A leaky bucket algorithm where requests enter a queue and are processed
    at a fixed rate. Requests are denied when the queue is at capacity.
    """

    def initial_state(self, config: RateLimitConfig) -> dict[str, Any]:
        """Return initial state with empty queue.

        Args:
            config: Rate limit configuration.

        Returns:
            Initial state with empty queue.
        """
        return {
            "queue_size": 0.0,
            "last_leak": 0.0,
        }

    def compute(
        self,
        state: dict[str, Any],
        config: RateLimitConfig,
        current_time: float,
    ) -> tuple[bool, dict[str, Any], dict[str, Any]]:
        """Compute whether a request is allowed and update state.

        Args:
            state: Current algorithm state dictionary.
            config: Rate limit configuration.
            current_time: Current Unix timestamp in seconds.

        Returns:
            Tuple of (allowed: bool, new_state: dict[str, Any], metadata: dict[str, Any]).
        """
        # Ensure pure function - do not mutate incoming state
        queue_size = state["queue_size"]
        last_leak = state["last_leak"]

        # Calculate capacity and leak rate
        capacity = config.burst_size if config.burst_size is not None else config.limit
        rate = config.tokens_per_second

        # Calculate elapsed time and drain the queue
        elapsed = max(0.0, current_time - last_leak)
        drained = elapsed * rate
        queue_after_leak = max(0.0, queue_size - drained)

        # Check if we can enqueue the new request
        if queue_after_leak + 1.0 <= capacity:
            # Allow request and enqueue it
            new_queue_size = queue_after_leak + 1.0
            allowed = True
            remaining = max(0, int(math.floor(capacity - new_queue_size)))
            metadata: dict[str, Any] = {"remaining": remaining}
        else:
            # Deny request - do not enqueue
            new_queue_size = queue_after_leak
            allowed = False

            # Calculate retry_after (guard against rate <= 0)
            if rate > 0:
                retry_after = math.ceil(
                    ((queue_after_leak + 1.0) - capacity) / rate
                )
                retry_after = max(1, retry_after)  # Ensure at least 1 second
            else:
                retry_after = 1
            remaining = max(0, int(math.floor(capacity - new_queue_size)))
            metadata = {"remaining": remaining, "retry_after": retry_after}

        # Create new state (pure function - don't mutate input)
        new_state: dict[str, Any] = {
            "queue_size": new_queue_size,
            "last_leak": current_time,
        }

        return allowed, new_state, metadata
